In [18]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import LeakyReLU, PReLU, ELU, ReLU
from tensorflow.keras.layers import Dropout
import warnings
warnings.filterwarnings("ignore")

In [3]:
data = pd.read_csv('D:/Final Projects/PRCP-1010-InsClaimPred/Data/train.csv')

In [4]:
data.head()

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
0   7       0          2              2          5              1   
1   9       0          1              1          7              0   
2  13       0          5              4          9              1   
3  16       0          0              1          2              0   
4  17       0          0              2          0              1   

   ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  ...  \
0              0              0              1              0  ...   
1              0              0              0              1  ...   
2              0              0              0              1  ...   
3              0              1              0              0  ...   
4              0              1              0              0  ...   

   ps_calc_11  ps_calc_12  ps_calc_13  ps_calc_14  ps_calc_15_bin  \
0           9           1           5           8               0   
1           3           1           1           9               0   
2           4           2           7           7               0   
3           2           2           4           9               0   
4           3           1           1           3               0   

   ps_calc_16_bin  ps_calc_17_bin  ps_calc_18_bin  ps_calc_19_bin  \
0               1               1               0               0   
1               1               1               0               1   
2               1               1               0               1   
3               0               0               0               0   
4               0               0               1               1   

   ps_calc_20_bin  
0               1  
1               0  
2               0  
3               0  
4               0  

[5 rows x 59 columns]

In [5]:
data.drop('id', inplace=True, axis=1)

In [6]:
data.head()

target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  ps_ind_05_cat  \
0       0          2              2          5              1              0   
1       0          1              1          7              0              0   
2       0          5              4          9              1              0   
3       0          0              1          2              0              0   
4       0          0              2          0              1              0   

   ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  ps_ind_09_bin  ...  \
0              0              1              0              0  ...   
1              0              0              1              0  ...   
2              0              0              1              0  ...   
3              1              0              0              0  ...   
4              1              0              0              0  ...   

   ps_calc_11  ps_calc_12  ps_calc_13  ps_calc_14  ps_calc_15_bin  \
0           9           1           5           8               0   
1           3           1           1           9               0   
2           4           2           7           7               0   
3           2           2           4           9               0   
4           3           1           1           3               0   

   ps_calc_16_bin  ps_calc_17_bin  ps_calc_18_bin  ps_calc_19_bin  \
0               1               1               0               0   
1               1               1               0               1   
2               1               1               0               1   
3               0               0               0               0   
4               0               0               1               1   

   ps_calc_20_bin  
0               1  
1               0  
2               0  
3               0  
4               0  

[5 rows x 58 columns]

In [7]:
x = data.iloc[:,1:]

In [8]:
x

ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  ps_ind_05_cat  \
0               2              2          5              1              0   
1               1              1          7              0              0   
2               5              4          9              1              0   
3               0              1          2              0              0   
4               0              2          0              1              0   
...           ...            ...        ...            ...            ...   
595207          3              1         10              0              0   
595208          5              1          3              0              0   
595209          1              1         10              0              0   
595210          5              2          3              1              0   
595211          0              1          8              0              0   

        ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  ps_ind_09_bin  \
0                   0              1              0              0   
1                   0              0              1              0   
2                   0              0              1              0   
3                   1              0              0              0   
4                   1              0              0              0   
...               ...            ...            ...            ...   
595207              0              0              0              1   
595208              0              0              0              1   
595209              1              0              0              0   
595210              0              0              1              0   
595211              1              0              0              0   

        ps_ind_10_bin  ...  ps_calc_11  ps_calc_12  ps_calc_13  ps_calc_14  \
0                   0  ...           9           1           5           8   
1                   0  ...           3           1           1           9   
2                   0  ...           4           2           7           7   
3                   0  ...           2           2           4           9   
4                   0  ...           3           1           1           3   
...               ...  ...         ...         ...         ...         ...   
595207              0  ...           4           1           9           6   
595208              0  ...           4           1           3           8   
595209              0  ...           3           2           2           6   
595210              0  ...           4           1           4           2   
595211              0  ...           4           4           3           8   

        ps_calc_15_bin  ps_calc_16_bin  ps_calc_17_bin  ps_calc_18_bin  \
0                    0               1               1               0   
1                    0               1               1               0   
2                    0               1               1               0   
3                    0               0               0               0   
4                    0               0               0               1   
...                ...             ...             ...             ...   
595207               0               1               1               0   
595208               1               0               1               0   
595209               0               0               1               0   
595210               0               1               1               1   
595211               0               1               0               0   

        ps_calc_19_bin  ps_calc_20_bin  
0                    0               1  
1                    1               0  
2                    1               0  
3                    0               0  
4                    1               0  
...                ...             ...  
595207               1               1  
595208               1               1  
595209               0               0  
595210

In [9]:
y = data.target

In [10]:
y

0         0
1         0
2         0
3         0
4         0
         ..
595207    0
595208    0
595209    0
595210    0
595211    0
Name: target, Length: 595212, dtype: int64

In [11]:
y.shape

(595212,)

In [12]:
y.unique()

array([0, 1], dtype=int64)

In [13]:
## Splitting the data into train test split
from sklearn.model_selection import train_test_split

In [14]:
X_train, X_test, y_train, y_test = train_test_split(x,y, test_size=0.2, random_state = 10)

In [15]:
X_train

ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  ps_ind_05_cat  \
246118          4              2          5              0              0   
487842          1              1          3              1              0   
326968          1              1          6              0              6   
222549          2              2          0              0              0   
554990          1              2          2              0              0   
...           ...            ...        ...            ...            ...   
299741          1              1          2              0              0   
533660          5              1          8              1              0   
105595          7              2          7              1              0   
443712          0              1          4              0              0   
345353          5              1          6              0              6   

        ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  ps_ind_09_bin  \
246118              1              0              0              0   
487842              1              0              0              0   
326968              1              0              0              0   
222549              0              0              0              1   
554990              0              0              1              0   
...               ...            ...            ...            ...   
299741              1              0              0              0   
533660              0              0              1              0   
105595              0              0              1              0   
443712              1              0              0              0   
345353              0              0              0              1   

        ps_ind_10_bin  ...  ps_calc_11  ps_calc_12  ps_calc_13  ps_calc_14  \
246118              0  ...           6           2           5          16   
487842              0  ...           4           0           3           8   
326968              0  ...           5           1           1          11   
222549              0  ...           5           3           4           3   
554990              0  ...           8           0           1          11   
...               ...  ...         ...         ...         ...         ...   
299741              0  ...           4           1           3          10   
533660              0  ...           9           2           1           6   
105595              0  ...           7           2           2           9   
443712              0  ...           2           1           4           6   
345353              0  ...           3           3           6           7   

        ps_calc_15_bin  ps_calc_16_bin  ps_calc_17_bin  ps_calc_18_bin  \
246118               1               1               0               1   
487842               0               1               1               0   
326968               0               1               1               0   
222549               0               1               1               0   
554990               0               1               1               0   
...                ...             ...             ...             ...   
299741               0               0               0               0   
533660               0               1               0               0   
105595               0               1               0               0   
443712               0               0               1               0   
345353               1               1               1               0   

        ps_calc_19_bin  ps_calc_20_bin  
246118               1               0  
487842               0               0  
326968               1               0  
222549               0               0  
554990               1               0  
...                ...             ...  
299741               1               0  
533660               0               0  
105595               0               0  
443712

In [16]:
X_train.shape

(476169, 57)

# Model Creation (ANN)

In [17]:
classifier = Sequential()
# Add the input layer
classifier.add(Dense(units=57, activation = 'relu'))
#add the first hidden layer
classifier.add(Dense(units = 7, activation = 'relu'))
#add the 2nd hidden layer
classifier.add(Dense(units=6, activation = 'relu'))
#adding the output layer
classifier.add(Dense(1, activation = 'sigmoid'))

In [20]:
classifier.compile(optimizer = opt , loss = 'binary_crossentropy', metrics = ['accuracy'])

In [21]:
import tensorflow

In [22]:
opt = tensorflow.keras.optimizers.Adam(learning_rate = 0.01)

In [39]:
model_history = classifier.fit(X_train, y_train, validation_split = 0.33, batch_size = 32, epochs = 10)

Epoch 1/10
9970/9970 [==============================] - 45s 5ms/step - loss: 0.1568 - accuracy: 0.9635 - val_loss: 0.1546 - val_accuracy: 0.9641
Epoch 2/10
9970/9970 [==============================] - 43s 4ms/step - loss: 0.1568 - accuracy: 0.9635 - val_loss: 0.1546 - val_accuracy: 0.9641
Epoch 3/10
9970/9970 [==============================] - 45s 4ms/step - loss: 0.1568 - accuracy: 0.9635 - val_loss: 0.1545 - val_accuracy: 0.9641
Epoch 4/10
9970/9970 [==============================] - 43s 4ms/step - loss: 0.1568 - accuracy: 0.9635 - val_loss: 0.1548 - val_accuracy: 0.9641
Epoch 5/10
9970/9970 [==============================] - 45s 4ms/step - loss: 0.1568 - accuracy: 0.9635 - val_loss: 0.1546 - val_accuracy: 0.9641
Epoch 6/10
9970/9970 [==============================] - 45s 4ms/step - loss: 0.1568 - accuracy: 0.9635 - val_loss: 0.1547 - val_accuracy: 0.9641
Epoch 7/10
9970/9970 [==============================] - 44s 4ms/step - loss: 0.1568 - accuracy: 0.9635 - val_loss: 0.1546 - val_ac

In [40]:
model_history.history.keys()

dict_keys(['loss', 'accuracy', 'val_loss', 'val_accuracy'])

In [43]:
# making predictions and evaluating the model
y_pred = classifier.predict(X_test)
y_pred = (y_pred >= 0.5)

3721/3721 [==============================] - 8s 2ms/step


# Model Evaluation

In [44]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
cm

array([[114633,      0],
       [  4410,      0]], dtype=int64)

In [45]:
from sklearn.metrics import accuracy_score
score = accuracy_score(y_pred, y_test)
score

0.9629545626370303

# Using Logistic Regression

In [19]:
## Step 3 Model creation
from sklearn.linear_model import LogisticRegression
clf=LogisticRegression()
clf.fit(X_train,y_train)

LogisticRegression()

In [20]:
y_hat=clf.predict(X_test)

In [21]:
y_hat

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [25]:
y_test

212793    0
228988    0
463972    0
210262    0
408929    0
         ..
418911    0
149481    0
505961    0
584045    0
43864     0
Name: target, Length: 119043, dtype: int64

In [22]:
from sklearn.metrics import confusion_matrix,accuracy_score,recall_score,precision_score,classification_report,f1_score

In [24]:
cr=classification_report(y_test,y_hat)
print(cr)

              precision    recall  f1-score   support

           0       0.96      1.00      0.98    114633
           1       0.00      0.00      0.00      4410

    accuracy                           0.96    119043
   macro avg       0.48      0.50      0.49    119043
weighted avg       0.93      0.96      0.94    119043



### Conclusion : 
This dataset does not contain any Column Names, so we don't have any visibility on what column is what type of data. Therefore we tried direct modelling with ANN and Logistic Regression and the accuracy is 96%. We can use either of the above model as it is giving the same accuracy.